File: java/mohammed-review/binarytree.java

class Node{
    Integer value;
    Node left;
    Node right;

    Node(int value){
        this.value = value;
        left = null;
        right = null;
    }
}

public class binarytree{
    Node root;

    private Node add(Node current, Integer value){
        if (current == null){
            return new Node(value);
        }

        if (value < current.value) {
            current.left = add(current.left, value);
        } else if (value > current.value) {
            current.right = add(current.right, value);
        } else{
            return current;
        }

        return current;
    }

    public void addNode(Integer i){
        root = add(root, i);
    }

    private binarytree createBinaryTree() {
        binarytree bt = new binarytree();
        return binarytree;
    }

    private boolean containsNode(Node current, int value) {
        if (current == null) {
            return false;
        }

        if (value == current.value) {
            return true;
        } 

        return value < current.value ? containsNode(current.left, value) : containsNode(current.right, value);

    }

    public boolean containsNode(int value) {
        return containsNode(root, value);
    }

    private Node deleteRecursive(Node current, int value) {
        if (current == null) {
            return null;
        }

        if (value == current.value) {
            if (current.left == null && current.right == null) {
                return null;
            } 

            if (current.right == null) {
                return current.left;
            }

            if (current.left == null) {
                return current.right;
            }

            int smallest = findSmallest(current.right);
            current.value = smallest;
            current.right = deleteRecursive(current.right, smallest);
            return current;
        }

        if (value < current.value) {
            current.left = deleteRecursive(current.left, value);
            return current;
        } else { 
            current.right = deleteRecursive(current.right, value);
            return current;
        }
    }

    private int findSmallest(Node root) {
        return root.left == null ? root.value : findSmallest(root.left);
    }

    public void delete(int value) {
        root = deleteRecursive(root, value);
    }


}

File: binarytree.py
class Node:

    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data

    def insert(self, item):
        if self.data:
            if item < self.data:
                if self.left is None:
                    self.left = Node(item)
                else:
                    self.left.insert(item)
            elif item > self.data:
                if self.right is None:
                    self.right = Node(item)
                else:
                    self.right.insert(item)
        else:
            self.data = item

    def print_tree(self):
        if self.left:
            self.left.print_tree()
        print(self.data)
        if self.right:
            self.right.print_tree()

root = Node(5)
root.insert(1)
root.insert(10)
root.insert(15)
root.print_tree()
                

